### ISSUE TITLE
IteratorResult doesn't have different types for yield values and return value

### ISSUE THREAD

**Issue by LinusU** [Microsoft/TypeScript#25558]
**TypeScript Version:**  master

**Search Terms:** IteratorResult, TReturn, TYield, Iterator, Iterable, IterableIterator

**Code**

```ts
function test (): IterableIterator<number> {
  let i = 0

  return {
    next (): IteratorResult<number> {
      if (i++ === 0) return { value: 0, done: false }
      if (i++ === 1) return { value: 1, done: false }
      if (i++ === 2) return { value: 2, done: false }

      return { value: undefined, done: true }
    },
    [Symbol.iterator] (): IterableIterator<number> {
      return this
    }
  }
}
```

**Expected behavior:** No errors since I'm returning an iterator that will only yield numbers

**Actual behavior:** The _return_ value of the function (nothing to do with the yielded values) is forced into a number as well, making the last return statement invalid.

**Playground Link:** [link](https://www.typescriptlang.org/play/#src=function%20test%20()%3A%20IterableIterator%3Cnumber%3E%20%7B%0D%0A%20%20let%20i%20%3D%200%0D%0A%0D%0A%20%20return%20%7B%0D%0A%20%20%20%20next%20()%3A%20IteratorResult%3Cnumber%3E%20%7B%0D%0A%20%20%20%20%20%20if%20(i%2B%2B%20%3D%3D%3D%200)%20return%20%7B%20value%3A%200%2C%20done%3A%20false%20%7D%0D%0A%20%20%20%20%20%20if%20(i%2B%2B%20%3D%3D%3D%201)%20return%20%7B%20value%3A%201%2C%20done%3A%20false%20%7D%0D%0A%20%20%20%20%20%20if%20(i%2B%2B%20%3D%3D%3D%202)%20return%20%7B%20value%3A%202%2C%20done%3A%20false%20%7D%0D%0A%0D%0A%20%20%20%20%20%20return%20%7B%20value%3A%20undefined%2C%20done%3A%20true%20%7D%0D%0A%20%20%20%20%7D%2C%0D%0A%20%20%20%20%5BSymbol.iterator%5D%20()%3A%20IterableIterator%3Cnumber%3E%20%7B%0D%0A%20%20%20%20%20%20return%20this%0D%0A%20%20%20%20%7D%0D%0A%20%20%7D%0D%0A%7D) **turn on strict null checks**

**Related Issues:** #23537

-----

This was mentioned by @rbuckton in #23537:

> In the long term, we need to consider a more "correct" version of `Iterator` and `IteratorResult` that keeps the `yield` and `return` types distinct, something like this:
>
> ```typescript
> interface Iterator<TYield, TReturn = TYield> {
>   next(value?: any): IteratorResult<TYield, TReturn>;
>   throw?(value?: any): IteratorResult<TYield, TReturn>;
>   return?(value: TReturn): IteratorResult<TYield, TReturn>;
> }
> interface IteratorYieldResult<TYield> {
>   value: TYield;
>   done: false;
> }
> interface IteratorReturnResult<TReturn> {
>   value: TReturn;
>   done: true;
> }
> type IteratorResult<TYield, TReturn = TYield> = IteratorYieldResult<TYield> | IteratorReturnResult<TReturn>;
> ```
>
> Then we could treat the yield and return types distinctly. Unfortunately, such a type would be a considerable breaking change.

I'm not sure I understand exactly why it would be a breaking change, since the ` = TYield` part should opt in all old code into the old behaviour, but I'm probably missing something.

-----

This problem also manifests itself with the inferred return types from generator functions:

```typescript
function * test () {
  yield 1 as number
  yield 2 as number
  return 'hej'
}

for (const i of test()) {
  i // i is typed as `number | string`, but it will **always** be a `number`
}
```

-----

My real world use case which led me to discover this issue was when manually creating an iterator from an IDBCursor (IndexedDB API):

```typescript
function createCursorIterator<Cursor extends IDBCursor, Item> (cursor: IDBRequest, mapFn: (cursor: Cursor) => Item): AsyncIterableIterator<Item> {
  let buffer: IteratorResult<Item>[] = []
  let queue: ((result: IteratorResult<Item>) => void)[] = []

  cursor.onsuccess = function () {
    const result: IteratorResult<Item> = this.result
      ? { value: mapFn(this.result), done: false }
      : { value: undefined, done: true }

    if (queue.length) {
      queue.shift()!(result)
    } else {
      buffer.push(result)
    }

    if (this.result) this.result.continue()
  }

  return {
    next (value?: any): Promise<IteratorResult<Item>> {
      return new Promise((resolve) => buffer.length ? resolve(buffer.shift()) : queue.push(resolve))
    },
    [Symbol.asyncIterator] (): AsyncIterableIterator<Item> {
      return this
    }
  }
}
```

The compiler doesn't allow me to assign `{ value: undefined, done: true }` to `result`.

**Comment by mhegazy** [comment]
Duplicate of https://github.com/Microsoft/TypeScript/issues/11375
